crates/turborepo-filewatch/src/hash_watcher.rs

```diff
@@ -0,0 +1,847 @@
+    scm: SCM,
+}
+
+enum Query {
+    GetHash(HashSpec, oneshot::Sender<Result<GitHashes, Error>>),
+}
+
+// Version is a type that exists to stamp an asynchronous hash computation with
+// a version so that we can ignore completion of outdated hash computations.
+#[derive(Clone, Default)]
+struct Version(Arc<()>);
+
+impl PartialEq for Version {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+}
+
+impl Eq for Version {}
+
+enum HashState {
+    Hashes(GitHashes),
+    Pending(Version, Vec<oneshot::Sender<Result<GitHashes, Error>>>),
+    Unavailable(String),
+}
+// We use a radix_trie to store hashes so that we can quickly match a file path
+// to a package without having to iterate over the set of all packages. We
+// expect file changes to be the highest volume of events that this service
+// handles, so we want to ensure we're efficient in deciding if a given change
+// is relevant or not.
+//
+// Our Trie keys off of a String because of the orphan rule. Keys are required
+// to be TrieKey, but this crate doesn't own TrieKey or AnchoredSystemPathBuf.
+// We *could* implement TrieKey in AnchoredSystemPathBuf and avoid the String
+// conversion, if we decide we want to add the radix_trie dependency to
+// turbopath.
+struct FileHashes(Trie<String, HashMap<Option<GlobSet>, HashState>>);
+
+impl FileHashes {
+    fn new() -> Self {
+        Self(Trie::new())
+    }
+
+    fn drop_matching<F>(&mut self, mut f: F, reason: &str)
+    where
+        F: FnMut(&AnchoredSystemPath) -> bool,
+    {
+        let mut previous = std::mem::take(&mut self.0);
+
+        // radix_trie doesn't have an into_iter() implementation, so we have a slightly
+        // inefficient method for removing matching values. Fortunately, we only
+        // need to do this when the package layout changes. It's O(n) in the
+        // number of packages, on top of the trie internals.
+        let keys = previous.keys().map(|k| k.to_owned()).collect::<Vec<_>>();
+        for key in keys {
+            let previous_value = previous
+                .remove(&key)
+                .expect("this key was pulled from previous");
+            let path_key =
+                AnchoredSystemPath::new(&key).expect("keys are valid AnchoredSystemPaths");
+            if !f(path_key) {
+                // keep it, we didn't match the key.
+                self.0.insert(key, previous_value);
+            } else {
+                for state in previous_value.into_values() {
+                    if let HashState::Pending(_, txs) = state {
+                        for tx in txs {
+                            let _ = tx.send(Err(Error::Unavailable(reason.to_string())));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn get_package_path(&self, file_path: &AnchoredSystemPath) -> Option<&AnchoredSystemPath> {
+        self.0
+            .get_ancestor(file_path.as_str())
```

Comment:
Is the ancestor here always going to be the parent directory? What if we have two files `apps/web/foo` and `apps/web/foobar`? Will the ancestor for the latter be `apps/web` or `apps/web/foo`? Could we add a test confirming this behavior?

Reply:
Good call on the test. I'm also adding some documentation, since this comment reflects a slight misunderstanding: the nodes in the tree are the package roots. If `apps/web` is a package, we would never have `apps/web/foo` and `apps/web/foobar`.

However, there is a case that I wasn't catching, which has now been fixed: if we have `apps/foo`, and someone modifies `apps/foodecoy`, _and it is not itself a package_, then this would return `apps/foo`, giving us a false positive package change.